content/courses/ada-idioms/chapters/component_access_to_rec_objs.rst

```diff
@@ -0,0 +1,915 @@
+and protected types can be used to declare record components. But there's more
+to this idiom.
+
+We would want a task type or protected type record component when:
+
+    a) a task or protected object (PO) is required to implement part |mdash|
+       but not all |mdash| of the record type's functionality, and
+
+    b) each such task or PO is intended to implement its functionality only for
+       the object logically containing that specific task object or protected
+       object. The record object and contained task/PO object pair is a
+       functional unit, independent of all other such units.
+
+This idiom applies to both enclosed task types and protected types, but for the
+sake of concision let's assume the record component will be of a protected
+type.
+
+As part of a functional unit, the PO component will almost certainly be
+required to reference the other record components in the enclosing record
+object. That reference will allow the PO to read and/or update those other
+components.  Note that these record components include discriminants, if any.
+
+To be a functional unit, the record object referenced by a given PO in this
+relationship must be the same record object at run-time that contains that
+specific PO instance. That will allow the PO instance to implement the
+functionality for the specific record object containing that PO instance.
+
+Unless we arrange it, that back-reference from the protected object to the
+record object isn't provided. Consider the following:
+
+.. code-block:: ada
+
+    package Q is
+       protected type P is ... end P;
+       type R is record
+         ...
+         Y : P;
+       end record;
+    end Q;
+
+During execution, whenever an object of type :ada:`Q.R` is declared or
+allocated, at run-time we will have two objects, instances of two distinct
+types |mdash| the record object and the protected object. Let's say that a
+client declares an object :ada:`Obj` of type :ada:`R`. There is only one
+reference direction defined, from the record denoted by :ada:`Obj` to the
+component protected object denoted by :ada:`Obj.Y`. This idiom, however,
+requires a reference in the opposite direction, from :ada:`Oby.Y` to
+:ada:`Obj`.
+
+This may seem like an unrealistic situation, but it is not. An IO device type
+that involves interrupt handling is just one real-world example, one that we
+will show in detail.
+
+The idiom context is a type because there will often be multiple real-world
+entities being represented in software. Representing these entities as multiple
+objects declared of a single type is by far the most reasonable approach.
+
+We assume the functional unit will be implemented as an
+:ref:`Abstract Data Type (ADT) <Ada_Idioms_Abstract_Data_Types>`. Strictly
+speaking, the ADT idiom is not required here, but that is the best approach for
+defining for types, for the good reasons given in that idiom entry. There's no
```

Comment:
"for defining for types" should be "for defining major types"
and
"for the good reasons given " should be "for the reasons given"

Reply:
Handled in #967.